Port over evil-magit

jojojames opened the pull request:
@justbur @Ambrevar @noctuid 

Quickly pulled over the evil-magit files, changed the namespace, and made a few tweaks around initialization.

Starting it up without the original evil-magit seems to work. README and makefile/testing could probably use some attention since I didn't try to run the test. It'd be nice if we could just get one "test" command to run through all our tests (well we only have evil-magit right now).

jojojames commented on Makefile:
```diff
@@ -20,4 +20,8 @@ lint:
 test:
 	cask exec ert-runner
 
+magit-test: elpa
```
Pulled this over from evil-magit, ideally, we'd be able to move this under test: somehow.

justbur replied:
Seems like there should be a discussion about how to run tests for each package generally.

jojojames replied:
I don't really have any thoughts on this. Open to thoughts from others.

jojojames replied:
I tweaked the test targets a little, magit-test will run magit tests and test: will run the only two tests we have now. Not the most polished but good enough, probably..

When I ran the tests, I did end up getting an error but I get the same error in the original evil-magit repo too.

```
Test evil-magit-section-map-tests condition:
    (ert-test-failed
     ((should
       (eq
	(lookup-key ... ...)
	(nth 2 binding)))
      :form
      (eq magit-diff-visit-worktree-file magit-diff-visit-file-worktree)
      :value nil))
   FAILED  4/5  evil-magit-section-map-tests
   passed  5/5  evil-magit-section-maps-accounted-for

Ran 5 tests, 4 results as expected, 1 unexpected (2019-10-20 16:02:10-0700)

1 unexpected results:
   FAILED  evil-magit-section-map-tests

make: *** [test] Error 1```

jojojames commented on modes/magit/evil-collection-magit.el:
```diff
@@ -1,36 +1,52 @@
+;; This library configures Magit and Evil to play well with each
+;; other. For some background see https://github.com/magit/evil-magit/issues/1.
+
+;; Installation and Use
```
This could probably use some attention, I removed some stuff but the doc is not totally appropriate now that it's in evil-collection.

justbur replied:
I would just keep the information in the readme. It's not very accessible here anyway

jojojames replied:
Sounds good, removed.

jojojames commented on modes/magit/evil-collection-magit.el:
```diff
@@ -39,13 +55,633 @@
+  (interactive)
+  (when (region-active-p) (deactivate-mark)))
+
+(defun evil-collection-magit-define-key (state map-sym key def)
```
I didn't look at the delta between this and evil-collection-define-key but I'm guessing evil-collection-define-key is needed for the key translation, definition, etc code.

justbur replied:
I think we can throw away the one from evil-magit and use the evil-collection one

jojojames replied:
Sounds good, removed.

jojojames commented on modes/magit/evil-collection-magit.el:
```diff
@@ -1,36 +1,52 @@
-;; Author: James Nguyen <james@jojojames.com>
-;; Maintainer: James Nguyen <james@jojojames.com>
+;; Author: Justin Burkett <justin@burkett.cc>
+;; Maintainer: Justin Burkett <justin@burkett.cc>
```
Double checking you have commit access to evil-collection, I would've guessed members of emacs-evil have it.

justbur replied:
Yes, I think so

Ambrevar commented on modes/magit/README.org:
```diff
@@ -0,0 +1,215 @@
+
+Everything is contained in evil-collection-magit.el, so you may download and load that file
+directly. The recommended method is to use MELPA via package.el (=M-x
+package-install RET evil-collection-magit RET=).
```
evil-collection-magit?

jojojames replied:
I just removed the section. It should be installed through evil-collection anyways.

Ambrevar commented on modes/magit/README.org:
```diff
@@ -0,0 +1,215 @@
+(require 'evil-collection-magit)
+#+END_SRC emacs-lisp
+
+** Note on evil usage
```
Shouldn't Evil be capitalized?  Same for the entire readme.org.

Ambrevar replied:
And for Magit.

jojojames replied:
I changed that one but too lazy overall to find those changes everwhere. :D

Ambrevar commented on modes/magit/README.org:
```diff
@@ -0,0 +1,215 @@
+
+Some may want =?= to search backward instead of launching the popup which is
+also bound to =h=. To get this behavior, add the following line after =(require
+'evil-collection-magit)= in your configuration.
```
What if we just require 'evil-collection?

jojojames replied:
Changed it to use evil-collection-init.

Ambrevar commented on modes/magit/README.org:
```diff
@@ -0,0 +1,215 @@
+
+Typically, to prevent evil-collection-magit from overriding the default behavior with evil
+and magit loaded, you should bind the respective key to =nil= after loading
+evil-collection-magit. For example, to make =escape= behave as default
```
Maybe a colon after "default".

jojojames replied:
Done.

Ambrevar commented on modes/magit/README.org:
```diff
@@ -0,0 +1,215 @@
+bindings and there is no elegant solution to this problem in my opinion. Text
+mode is the best that I have come up with.
+
+* Key Bindings
```
No titlecase for consistency with other sections.

jojojames replied:
I think it's already consistent unless I'm misunderstanding.

Ambrevar commented on modes/magit/README.org:
```diff
@@ -0,0 +1,215 @@
+(evil-define-key* evil-collection-magit-state magit-mode-map [escape] nil)
+#+END_SRC
+
+* Known Conflicts
```
No titlecase for consistency with other sections.

jojojames replied:
Should be consistent already.

Ambrevar replied:
I've looked again, and actually many more sections than those 2 are titlecased (or partially).  But many are sentence-cased.

Ambrevar commented on modes/magit/evil-collection-magit.el:
```diff
@@ -1,36 +1,52 @@
-;;; evil-collection-magit.el --- Bindings for `magit' -*- lexical-binding: t -*-
+;;; evil-collection-magit.el --- evil-based key bindings for magit
```
Should lexical bindings really be off?

jojojames replied:
The original was off, I believe. I don't want to set it to lexical binding and have to make code changes.

Ambrevar commented:
Neat, thanks for the hard work you all!

justbur commented:
What is the plan for transitioning people?

jojojames commented:
@justbur I think we can add a notice to the original repo and play it out similar to the other merges like evil-ediff/evil-mu4e.

I think there will be some errors if you have both evil-magit and evil-collection-magit setup at the same time. Not too sure if it's worth expending some effort to do some compatibility checks.

Silex commented:
Can you explain what the general idea of the PR is? replace evil-magit?

jojojames commented:
@Silex I think the linked threads will provided better context. In short, we're trying to do to evil-magit what we did to evil-ediff.

Silex commented:
```
evil-ediff

This project has been merged into the larger Evil Collection project and will no longer be maintained here.
```

Okay so basically, merge evil-magit here. Thanks :+1:

ethancedwards8 commented:
What's the status of this getting merged any time soon?

jojojames commented:
Most of the work had been done a while back but we also ported over some of the failing tests. I think with a few tweaks, it would be good to go.

ethancedwards8 commented:
@jojojames not to bug you but, I would love a status update on this?

jojojames commented:
@ethancedwards8 We'd need to update this patch to the latest evil-magit changes (shouldn't be to hard honestly) and then with @justbur approval, we can merge this. We'd then need to update the evil-magit repo with a notice to point users to evil-collection.

I'm happy to have someone else push this to its end as I'm a little busy recently to push this forward.

justbur commented:
I approve, and will happily (I have very little extra time right now) put the note in the README when this happens.

jojojames commented:
@justbur I've updated this PR with the latest changes from evil-magit and fixed the one test failure from a year ago.

Waiting on your final approval before merging.

ethancedwards8 commented:
I'm glad to see this is finally happening! Seems like its been a long thing coming.

justbur commented:
@jojojames Go ahead and merge when you're ready. I'll update the readme at evil-magit and probably archive the repo at some point

jojojames commented:
Thanks, merged.
